File: solutions/python/luhn/1/luhn.py
import re
import functools


class Luhn:
    def __init__(self, card_num: str):
        self.card_num = card_num

    def valid(self) -> bool:
        if re.match(r'^[\d|\s]+$', self.card_num) is None:
            return False

        *card_num, = map(int, filter(str.isdigit, self.card_num))

        if len(card_num) <= 1:
            return False

        card_num = card_num[::-1]

        def luhn_sum(acc: int, value: tuple[int, int]) -> int:
            index, number = value

            def luhn_doubling(number: int) -> int:
                number *= 2
                return number if number < 10 else number - 9

            return acc + (number if index % 2 == 0 else luhn_doubling(number))

        return functools.reduce(luhn_sum, enumerate(card_num), 0) % 10 == 0

File: solutions/ruby/luhn/1/luhn.rb
module Luhn
  def self.valid?(num)
    num = num.gsub(/\s/,'')
    return false if num.length <= 1 or num.match(/[^\d]/)
    num.chars
        .map(&:to_i)
        .reverse
        .map
        .with_index {|n, index| index % 2 == 0 ? n : n * 2}
        .map {|n| n > 9 ? n - 9 : n}
        .sum % 10 == 0
  end
end